src/main/java/picard/nio/PicardBucketUtils.java

```diff
@@ -1,59 +1,57 @@
 package picard.nio;
 
 import com.google.cloud.storage.contrib.nio.CloudStorageFileSystem;
+import com.google.cloud.storage.contrib.nio.CloudStoragePath;
+import htsjdk.io.IOPath;
 import htsjdk.samtools.util.FileExtensions;
 
 import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.Arrays;
 import java.util.UUID;
 
 
 /**
- * Copied from BucketUtils.java in GATK
- * To be replaced once the original GATK BucketUtils.java is ported to htsjdk
+ * Derived from BucketUtils.java in GATK
  */
-public class GATKBucketUtils {
-    // In GATK these are accessed as e.g. GoogleCloudStorageFileSystem.SCHEME
+public class PicardBucketUtils {
     public static final String GOOGLE_CLOUD_STORAGE_FILESYSTEM_SCHEME = "gs";
     public static final String HTTP_FILESYSTEM_PROVIDER_SCHEME = "http";
     public static final String HTTPS_FILESYSTEM_PROVIDER_SCHEME = "https";
-
-
-    public static final String GCS_PREFIX = "gs://";
-    public static final String HTTP_PREFIX = "http://";
-    public static final String HTTPS_PREFIX = "https://";
     public static final String HDFS_SCHEME = "hdfs";
-    public static final String HDFS_PREFIX = HDFS_SCHEME + "://";
+    public final static String FILE_SCHEME = "file";
 
     // slashes omitted since hdfs paths seem to only have 1 slash which would be weirder to include than no slashes
-    public static final String FILE_PREFIX = "file:";
-
-    private GATKBucketUtils(){} //private so that no one will instantiate this class
+    private PicardBucketUtils(){} //private so that no one will instantiate this class
 
     /**
      * Get a temporary file path based on the prefix and extension provided.
-     * This file (and possible indexes associated with it) will be scheduled for deletion on shutdown
+     * This file (and possible indexes associated with it) will be scheduled for deletion on shutdown.
+     *
+     * As this is an "initial" path operation, it is OK to take for the prefix and extension variables to be strings.
      *
      * @param prefix a prefix for the file name
      *               for remote paths this should be a valid URI to root the temporary file in (e.g. gs://hellbender/staging/)
      *               there is no guarantee that this will be used as the root of the tmp file name, a local prefix may be placed in the tmp folder for example
-     * @param extension and extension for the temporary file path, the resulting path will end in this
+     * @param extension and extension for the temporary file path, the resulting path will end in this. It should include the period
+     *                  e.g. ".txt" extension of "txt" results in a filename "prefixtxt" without a period.
      * @return a path to use as a temporary file, on remote file systems which don't support an atomic tmp file reservation a path is chosen with a long randomized name
```

Comment:
```suggestion
     * @return a path to use as a temporary file; if the prefix specifies a remote file system which doesn't support an atomic temporary file reservation operation, the temporary file will be located in a local path and use a long randomized name
```
Or something along those lines.

Reply:
incorporated

Reply:
Actually, I think the variable name being "prefix" is a little misleading. "prefix" is really a directory, and it always gives the new temp file a randomized name (at least that's the behavior for the GCS and local filesystem).

Reply:
(Update): now this method takes "directory", "prefix", and "extension" as arguments, like `randomRemotePath`